gtest/src/address.rs

```diff
@@ -0,0 +1,75 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use gear_core::program::ProgramId;
+use serde::{Deserialize, Deserializer, Serialize};
+use std::str::FromStr;
+
+use sp_core::{crypto::Ss58Codec, hexdisplay::AsBytesRef, sr25519::Public};
+use sp_keyring::sr25519::Keyring;
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(tag = "kind", content = "value")]
+pub enum Address {
+    #[serde(rename = "account")]
+    Account(String),
+    #[serde(rename = "id")]
+    ProgramId(u64),
+    #[serde(rename = "ss58")]
+    SS58(String),
+}
+
+impl Default for Address {
+    fn default() -> Self {
+        Self::Account("alice".to_string())
+    }
+}
+
+impl Address {
+    pub fn into_program_id(&self) -> ProgramId {
```

Comment:
Don't break naming conventions. Name it say `to_program_id(...)`.
https://rust-lang.github.io/api-guidelines/naming.html#c-conv

Reply:
What about to just impl `From` ?

Reply:
`ProgramId` lives in `gcore` crate/module, so impl `From` is not-trivial.
However `Into` could be implemented but again  - it is consuming too. Nevertheless `Into` could be implemented for `&Address` though the implementation doesn't bring any advantages: `message.destination.as_ref().into()` or `message.destination.borrow().into()` is required.

What is better, just rename the method?

Reply:
Up to you. Renaming seems well